rust/arrow/src/compute/kernels/concat.rs

```diff
@@ -0,0 +1,310 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Defines concat kernel for `ArrayRef`
+//!
+//! Example:
+//!
+//! ```
+//! use std::sync::Arc;
+//! use arrow::array::{ArrayRef, StringArray};
+//! use arrow::compute::concat;
+//!
+//! let arr = concat(&vec![
+//!     Arc::new(StringArray::from(vec!["hello", "world"])) as ArrayRef,
+//!     Arc::new(StringArray::from(vec!["!"])) as ArrayRef,
+//! ]).unwrap();
+//! assert_eq!(arr.len(), 3);
+//! ```
+
+use crate::array::*;
+use crate::datatypes::*;
+use crate::error::{ArrowError, Result};
+
+use TimeUnit::*;
+
+/// Concatenate multiple `ArrayRef` with the same type.
+///
+/// Returns a new ArrayRef.
+pub fn concat(array_list: &Vec<ArrayRef>) -> Result<ArrayRef> {
+    let mut data_type: Option<DataType> = None;
+    let array_data_list = &array_list
+        .iter()
+        .map(|a| {
+            let array_data = a.data_ref().clone();
+            let curr_data_type = array_data.data_type().clone();
+            match &data_type {
+                Some(t) => {
+                    if t != &curr_data_type {
+                        return Err(ArrowError::ComputeError(
+                            "Cannot concat arrays with different data types".to_string(),
+                        ));
+                    }
+                }
+                None => {
+                    data_type = Some(curr_data_type);
+                }
+            }
+            Ok(array_data)
+        })
+        .collect::<Result<Vec<ArrayDataRef>>>()?;
+
+    let data_type = match data_type {
```

Comment:
If you change this to check that `!array_list.is_empty()` at the start of the function, you could then get the datatype from the first element in `array_list`, then avoid validating data types as `append_data` will do that for us